File: Deepspace/lib/ShieldBooster.rb
#encoding:utf-8

module Deepspace

# Class to represent the shield boosters that space stations can have
class ShieldBooster

	# Constructors
	#===========================================================================
	
	# Description:
	# 	Class initializer
	# Parameters:
	# 	name: String, name of the shield booster
	# 	boost: Float, percentage of damage absorbed by the shield
	# 	uses: Integer, how many uses the ShieldBooster has
	# Return:
	# 	Nil
	def initialize(_name, _boost, _uses)
		@name = _name
		@boost = _boost
		@uses = _uses
	end
	
	# Description:
	# 	Copy Constructor
	# Parameters:
	# 	origin: ShieldBooster instance which is going to be copied
	# Return:
	# 	A ShieldBooster instance holding a copy of origin
	def self.newCopy(origin)
		return new(origin.name, origin.boost, origin.uses)
	end

	# Getters
	#===========================================================================

	attr_reader :boost, :uses, :name

	# Description:
	# 	Gets the UI representation of the ShieldBooster object
	# Returns:
	#	ShieldToUI: the UI representation
	def getUIVersion
		return ShieldToUI.new(self)
	end

	# Description:
	# 	Displays all the relevant data of the object
	# Returns:
	# 	String, containing all parameters value
	def to_s
		return "[ShieldBooster]-> Boost: #{@boost}, Uses: #{@uses}"
	end

	# Description:
	# 	Gets the UI representation of the object
	# Returns:
	# 	ShieldToUI: the UI representation
	def getUIVersion
		return ShieldToUI.new(self)
	end

	# Setters
	#===========================================================================

	# Description:
	# 	Uses the shield booster
	# 	If it is still available (uses is greater than 0), we can use the boost
	# 	Otherwise, we cannot use the boost
	# Return:
	# 	Float:
	# 		Boost if uses > 0
	#		1.0 if uses == 0
	def useIt
		if @uses > 0			# We can use our shield booster
			@uses -= 1
			return @boost
		else					# Shield booster is drained
			return 1.0
		end
	end

end # class ShieldBooster

end # module Deepspace

File: deepspace/src/ShieldBooster.java
/**
 * @author Sergio Quijano Rey
 * */

package Deepspace;

/**
 * @brief Class to represent shield booster that space stations can have
 * */
class ShieldBooster{
	// Private attributes
	//======================================================================
	private String name;
	private float boost;
	private int uses;

	// Constructors
	//======================================================================
	ShieldBooster(String _name, float _boost, int _uses){
		name = _name;
		boost = _boost;
		uses = _uses;
	}

	/**
	 * @brief Copy constructor
	 * @param other: the ShieldBooster to be copied
	 * @post the object has an exact copy of the other given object
	 * */
	ShieldBooster(ShieldBooster other){
		this(other.name, other.boost, other.uses);
	}

	// Getters
	//======================================================================
	public float getBoost(){
		return boost;
	}

	public int getUses(){
		return uses;
	}

	// Setters
	//======================================================================
	/**
	 * @brief method to use the shield booster
	 * @return 	boost, if there are uses left
	 * 	   		1.0 (neutral boost) otherwise
	 * */
	public float useIt(){
		if(uses > 0){
			uses = uses - 1;
			return boost;
		}else{
			return 1.0;
		}

		// Security check
		System.out.println("WARNING! Unexpected condition at ShieldBooster.useIt()");
		return 1.0;
	}
}
